src/measure_theory/independence.lean

```diff
@@ -0,0 +1,297 @@
+measurable spaces they generate: a set `s` generates the measurable space with measurable sets
+`∅, s, sᶜ, univ`.
+* Independence of functions (or random variables) is also defined as independence of the measurable
+spaces they generate : a function `f` for which we have a measurable space `m` on the codomain
+generates `measurable_space.comap f m`.
+
+## Main statements
+
+* TODO: `indep_of_indep_sets`: if π-systems are independent as sets of sets, then the
+measurable spaces they generate are independent.
+* `indep2_of_indep2_sets`: variant with two π-systems.
+
+## Implementation notes
+
+We provide one main definition of independence:
+* `indep_sets`: independence of a family of sets of sets `pi : ι → set (set α)`,
+Three other independence notions are defined using `indep_sets`:
+* `indep`: independence of a family of measurable spaces `m : ι → measurable_space α`,
+* `indep_set`: independence of a family of sets `s : ι → set α`,
+* `indep_fun`: independence of a family of functions. For measurable spaces
+  `m : Π (x : ι), measurable_space (β x)`, we consider functions `f : Π (x : ι), α → β x`.
+
+Additionally, we provide four corresponding statements for two measurable spaces (resp. sets of
+sets, sets, functions) instead of a family.
+
+The definition of independence for `indep_sets` uses finite sets (`finset`). An alternative and
+equivalent way of defining independence would have been to use countable sets.
+TODO: prove that equivalence.
+
+Most of the definitions and lemma in this file list all variables instead of using the `variables`
+keyword at the beginning of a section, for example
+`lemma indep2.symm {α} {m₁ m₂ : measurable_space α} [measurable_space α] {μ : measure α} ...` .
+This is intentional, to be able to control the order of the `measurable_space` variables. Indeed
+when defining `μ` in the example above, the measurable space used is the last one defined, here
+`[measurable_space α]`, and not `m₁` or `m₂`.
+
+## References
+
+* Williams, David. Probability with martingales. Cambridge university press, 1991.
+Part A, Chapter 4.
+-/
+
+open measure_theory measurable_space
+open_locale big_operators classical
+
+section definitions
+
+/-- A family of sets of sets `π : ι → set (set α)` is independent with respect to measure `μ` if for
+any finite set of indices `s = {i_1, ..., i_n}`, for any sets
+`f i_1 ∈ π i_1, ..., f i_n ∈ π i_n`, `μ (⋂ i in s, f i) = ∏ i in s, μ (f i) `.
+It will be used for families of pi_systems. -/
+def indep_sets {α ι} [measurable_space α] (π : ι → set (set α)) (μ : measure α) : Prop :=
+∀ (s : finset ι) {f : ι → set α} (H : ∀ i, i ∈ s → f i ∈ π i), μ (⋂ i ∈ s, f i) = ∏ i in s, μ (f i)
+
+/-- Two sets of sets `s₁, s₂` are independent with respect to
+measure `μ` if for any sets `t₁ ∈ p₁, t₂ ∈ s₂`, `μ (t₁ ∩ t₂) = μ (t₁) * μ (t₂)` -/
+def indep2_sets {α} [measurable_space α] (s1 s2 : set (set α)) (μ : measure α) : Prop :=
+∀ t1 t2 : set α, t1 ∈ s1 → t2 ∈ s2 → μ (t1 ∩ t2) = μ t1 * μ t2
```

Comment:
`indep` and `Indep` might be nice ways to distinguish the n-ary and binary definition, as we do with `Sup` / `sup` and `Union` / `union`.

Reply:
I love the idea. Thanks!